### PR TITLE
Added first / last buttons and max_lines to pagination

### PR DESCRIPTION
**Pull request by runew0lf**

### CODE REVIEW

**Comment by 1mn** [comment]
Did some pair programming with rune to refactor this.

- Made this into a subclass LinePaginator
- moved `paginate` from utils into a class method in LinePaginator
- updated all usages of the util `paginate` inside tags and clickup to work with the new subclass.
- updated the logic so that the max_char approach still works, and max_lines is entirely optional.

this is ready for real review now.

File: bot/pagination.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+            Make sure that this reaction is what we want to operate on
+            """
+
+            return (
+                reaction_.message.id == message.id and  # Reaction on this specific message
+                reaction_.emoji in PAGINATION_EMOJI and  # One of the reactions we handle
+                user_.id != ctx.bot.user.id and (  # Not applied by the bot itself
+                    not restrict_to_user or   # Unrestricted if there's no user to restrict to, or...
+                    user_.id == restrict_to_user.id  # Only by the restricted user
+                )
```
Comment (AnonGuy) [inline bot/pagination.py]: Could we have something a bit more readable for the conditions? Maybe something like:
```py
no_restrictions = (
    # Pagination is not restricted
    not restrict_to_user or
    # The reaction was by a whitelisted user
    user_.id == restrict_to_user.id
)

return (
    # Conditions for a successful pagination:
    all(
        # Reaction is on this message
        reaction_.message.id == message.id,
        # Reaction is one of the pagination emotes
        reaction_.emoji in PAGINATION_EMOJI,
        # Reaction was not made by the Bot
        user_.id != ctx.bot.user.id,
        # There were no restrictions
        no_restrictions
    )
)
```

File: bot/pagination.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+                all(
+                    # Reaction is on this message
+                    reaction_.message.id == message.id,
+                    # Reaction is one of the pagination emotes
+                    reaction_.emoji in PAGINATION_EMOJI,
+                    # Reaction was not made by the Bot
+                    user_.id != ctx.bot.user.id,
+                    # There were no restrictions
+                    no_restrictions
+                )
```
Comment (AnonGuy) [inline bot/pagination.py]: Looks like I dun goofed. You'll need an extra set of parens around the arguments for `all`:
```
no_restrictions = (
    # Pagination is not restricted
    not restrict_to_user or
    # The reaction was by a whitelisted user
    user_.id == restrict_to_user.id
)

return (
    # Conditions for a successful pagination:
    all((
        # Reaction is on this message
        reaction_.message.id == message.id,
        # Reaction is one of the pagination emotes
        reaction_.emoji in PAGINATION_EMOJI,
        # Reaction was not made by the Bot
        user_.id != ctx.bot.user.id,
        # There were no restrictions
        no_restrictions
    ))
)
```

**Comment by runew0lf** [comment]
Thanks buddy, had some flake8 issues, but theyre resolved now

**Comment by AnonGuy** [comment]
Alright, now that Travis has been sated, it looks like we're good to go!

**Pull request merged**